### PR TITLE
Small improvements to the OOP chapter

### PR DESCRIPTION
**Pull request by bernhardmgruber**
I read through the OOP chapter and here are some suggested changes. This PR is not intending to fix critical issues, but rather tries to improve the working, correctness and accuracy in some places. Sometimes I also just liked a different phrasing better. Feel free to comment on any of them!

### CODE REVIEW

File: talk/objectorientation/functors.tex [hunk 1]
```diff
@@ -36,13 +36,13 @@
...
-    };
     double binary_op(double a, double b, BinaryFunction &func) {
       return func(a, b);
     }
+    struct Add : BinaryFunction {
+      double operator() (double a, double b) override
+      { return a+b; }
+    };
     Add addfunc;
     double c = binary_op(a, b, addfunc);
```
Comment (bernhardmgruber) [inline talk/objectorientation/functors.tex:47]: I found it more logical to show that `binary_op` can be implemented without knowing `Add` yet.

File: talk/objectorientation/objectsclasses.tex [hunk 2]
```diff
@@ -13,7 +13,7 @@
   \begin{block}{Objects}
     instances of classes
   \end{block}
-  \begin{block}{A class encapsulates a concept}
+  \begin{block}{A class encapsulates state and behavior of ``something''}
```
Comment (bernhardmgruber) [inline talk/objectorientation/objectsclasses.tex:16]: I just disliked "concept" here. Feel free to suggest something better!

File: talk/objectorientation/constructors.tex [hunk 3]
```diff
@@ -84,9 +84,9 @@
   \begin{block}{Concept}
     \begin{itemize}
     \item special constructor called for replicating an object
-    \item takes a single parameter of type const ref to class
-    \item will be implemented by the compiler if not provided
-    \item in order to forbid copy, use {\texttt delete} (coming in 2 slides)
+    \item takes a single parameter of type \mintinline{cpp}{const &} to class
+    \item provided by the compiler if not by the user
+    \item in order to forbid copy, use \mintinline{cpp}{= delete} (coming in 2 slides)
```
Comment (hageboeck) [inline talk/objectorientation/constructors.tex:89]:
```suggestion
    \item in order to forbid copy, use \mintinline{cpp}{= delete} (see next slides)
```
Those kind of things go stale quickly.

File: talk/objectorientation/constructors.tex [hunk 4]
```diff
@@ -100,9 +100,9 @@
     };
   \end{cppcode}
   \pause
-  \begin{exampleblock}{The rule of 3/5 (\cpp98/\cpp11 and newer)}
+  \begin{exampleblock}{The rule of 3/5/0 (\cpp98/\cpp11 and newer) - {\color{blue!50!white} \href{https://en.cppreference.com/w/cpp/language/rule_of_three}{cppreference}}}
```
[the remark this answers is not in the document]
Comment (hageboeck) [reply talk/objectorientation/constructors.tex:103]: That's very possible in latex. It's in the setup of the hyperref package.

File: talk/objectorientation/constructors.tex [hunk 5]
```diff
@@ -122,24 +122,24 @@
...
-    \item by declaring them as default
+    \item avoid empty default constructors like \mintinline{cpp}{ClassName() {}}
+    \item declare them as \mintinline{cpp}{= default}
     \end{itemize}
   \end{block}
   \begin{block}{Details}
     \begin{itemize}
     \item when no user defined constructor, a default is provided
-    \item any user defined constructor disables default ones
+    \item any user defined constructor disables the default one
```
Comment (hageboeck) [inline talk/objectorientation/constructors.tex:142]:
```suggestion
    \item any user-defined constructor disables the default one
```

**Review by hageboeck — COMMENTED** [review]
In general very good proposals. I added some more + some refinements.

File: talk/objectorientation/constructors.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply talk/objectorientation/constructors.tex:103]: Sure, I know that :) But I did not dare to invade the presentation this much. I quickly grepped for `href` and most often links are represented as `beamergotobutton`, sometimes inconsistently colored. I guess this should be harmonized in a separate stage.

File: talk/objectorientation/adl.tex [hunk 6]
```diff
@@ -30,17 +30,17 @@
...
-    \item current class if any, base class if any, etc...
-    \item for names used to the left of the () operator, Argument Dependent Lookup (ADL)
+    \item current namespace/block, enclosing namespaces/blocks, etc...
+    \item current class if any, base classes if any, etc...
+    \item for a call expression (e.g. \mintinline{cpp}{f(a, b)} or \mintinline{cpp}{a + b}), Argument Dependent Lookup (ADL)
     \end{itemize}
   \end{block}
   \begin{exampleblock}{Argument Dependent Lookup (simplified)}
-    For each argument of a function call, the function name is also searched in:
+    The called name in a call expression (including operators) is additionally searched in each argument's:
```
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply talk/objectorientation/adl.tex]: Fixed.

File: talk/objectorientation/advancedoo.tex [hunk 7]
```diff
@@ -177,7 +177,7 @@
   \begin{alertblock}{Consequences}
     \begin{itemize}
     \item virtual methods are ``slower'' than standard ones
-    \item and they cannot be inlined
+    \item and they can seldom be inlined
```
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex]: Fixed.

File: talk/objectorientation/advancedoo.tex [hunk 8]
```diff
@@ -249,10 +250,10 @@
   \frametitlecpp[98]{Pure Virtual methods}
   \begin{block}{Concept}
     \begin{itemize}
-    \item methods that exist but are not implemented
-      \item marked by ``{\tt = 0} '' in the declaration
+    \item unimplemented methods that must be overridden
+      \item marked by \mintinline{cpp}{= 0} in the declaration
```
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex]: Fixed.

File: talk/objectorientation/constructors.tex [hunk 9]
```diff
@@ -84,9 +84,9 @@
   \begin{block}{Concept}
     \begin{itemize}
     \item special constructor called for replicating an object
-    \item takes a single parameter of type const ref to class
-    \item will be implemented by the compiler if not provided
-    \item in order to forbid copy, use {\texttt delete} (coming in 2 slides)
+    \item takes a single parameter of type \mintinline{cpp}{const &} to class
+    \item provided by the compiler if not by the user
```
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply talk/objectorientation/constructors.tex]: Fixed.

File: talk/objectorientation/constructors.tex (hunk 3 above)
Comment (bernhardmgruber) [reply talk/objectorientation/constructors.tex]: Yeah, they do. Fixed.

File: talk/objectorientation/constructors.tex (hunk 5 above)
Comment (bernhardmgruber) [reply talk/objectorientation/constructors.tex]: Fixed.

File: talk/objectorientation/constructors.tex (hunk 4 above)
[the remark this answers is not in the document]
Comment (hageboeck) [reply talk/objectorientation/constructors.tex:103]: I did it in 97. 🙂 
Will check the buttons after the CI finishes.

File: talk/objectorientation/advancedoo.tex [hunk 10]
```diff
@@ -27,8 +27,8 @@
       \classbox{Drawable}{}
       \classbox[below of=Drawable]{Shape}{}
       \classbox[below of=Shape]{Polygon}{}
-      \draw[very thick,->] (Polygon) -- (Shape);
-      \draw[very thick,->] (Shape) -- (Drawable);
+      \draw[very thick,-Triangle] (Polygon) -- (Shape);
+      \draw[very thick,-Triangle] (Shape) -- (Drawable);
```
[the remark this answers is not in the document]
Comment (sponce) [reply talk/objectorientation/advancedoo.tex:31]: :+1:

File: talk/objectorientation/advancedoo.tex [hunk 11]
```diff
@@ -476,6 +477,9 @@
 
 \begin{frame}[fragile]
   \frametitlecpp[98]{Virtual inheritance}
+  \begin{exampleblock}{Good practice}
+    if you write a class and expect users to inherit from it, declare its destructor \mintinline{cpp}{virtual}
+  \end{exampleblock}
```
[the remark this answers is not in the document]
Comment (sponce) [reply talk/objectorientation/advancedoo.tex:482]: Indeed we should have this somewhere, especially that we precisely show the consequences of such a mistake in one of the exercises (memory leak in that case).
Maybe we add a dedicated slide with a first point stating that destructors can also be virtual (not obvious for all I suppose) and what it means on which destructor is called.
Then your point would come as a natural conclusion, and I would just rephrase it to "if you write a class with at least one virtual method, declare its destructor \mintinline{cpp}{virtual}".

File: talk/objectorientation/objectsclasses.tex (hunk 2 above)
Comment (sponce) [reply talk/objectorientation/objectsclasses.tex:16]: Indeed, now that concept has some other meaning in C++, it does not feel right anymore. I'm happy with "something" !

File: talk/objectorientation/advancedoo.tex [hunk 12]
```diff
@@ -177,7 +177,7 @@
   \begin{alertblock}{Consequences}
     \begin{itemize}
     \item virtual methods are ``slower'' than standard ones
-    \item and they cannot be inlined
+    \item and they can seldomly be inlined
```
[the remark this answers is not in the document]
Comment (sponce) [reply talk/objectorientation/advancedoo.tex]: Indeed, not sure we want to go to these details right here but new phrasing is definitely better.

**Comment by sponce** [comment]
Thanks for all these nice improvements ! I'll merge now so that I can benefit from them tomorrow morning (I'm giving the course at ESIPAP). You can then move/improve the good practice sentence in a next MR.

File: talk/objectorientation/advancedoo.tex (hunk 11 above)
[the remark this answers is not in the document]
Comment (bernhardmgruber) [reply talk/objectorientation/advancedoo.tex:482]: I created #99 for that.
